### Capstone Project Notebook: Week 3 Assignment: Segmenting and Clustering Neighborhoods in Toronto

In [2]:
import pandas as pd
import numpy as np
import requests

#### Using 'pandas' to parse wikipage containing postal codes of Canada.
#### Converting the table into a DataFrame called 'TorontoPostalCodes'.

In [3]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

html = requests.get(url).content
df_list = pd.read_html(html)
TorontoPostalCodes = df_list[0]
TorontoPostalCodes

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

#### Dropping all rows where 'Borough' value has 'Not Assigned'.
#### Resting index with new sequential index.

In [4]:
TorontoPostalCodes = TorontoPostalCodes[TorontoPostalCodes.Borough !='Not assigned']
TorontoPostalCodes = TorontoPostalCodes.reset_index(drop=True)
TorontoPostalCodes

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [27]:
TorontoPostalCodes.shape

(103, 3)

#### There are 103 postal codes (rows) with an assigned Borough and respective neighbourhoods listed against each postal code.

# End of Part 1 of 3

In [5]:
!pip install pgeocode
import pgeocode # import geocoder

In [29]:
TorontoPostalCodes.columns = TorontoPostalCodes.columns.str.replace(' ', '')
dfLL = pd.DataFrame(columns = ['Latitude','Longitude'])
nomi = pgeocode.Nominatim('ca')

for n in TorontoPostalCodes.index:
    df = nomi.query_postal_code(TorontoPostalCodes.PostalCode[n])
    data = {'Latitude': df.latitude, 'Longitude':df.longitude}
    dfLL = dfLL.append(data, ignore_index=True)

TorontoPostalCodes[['Latitude','Longitude']] = dfLL[['Latitude','Longitude']]
TorontoPostalCodes

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  Latitude  Longitude  
0                                            Parkwoods   43.7545   -79.3300  
1                                     Victoria Village   43.7276   -79.3148  
2                            Regent Park, Harbourfront   43.6555   -79.3626  
3                     Lawrence Manor, Lawrence Heights   43.7223   -79.4504  
4          Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
..                                                 ...       ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North   43.6518   -79.5076  
99                                Church and Wellesley   43.6656   -79.3830  
100  Business reply mail Processing Centre, South C...   43.7804   -79.2505  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...   43.6325   -79.4939  
102  Mimico NW, The Queensway West, South of Bloor,...   43.6256   -79.5231  

[103 rows x 5 columns]

# End of Part 2 of 3

In [7]:
import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium==0.5.0
import folium # map rendering library

print('Libraries imported.')

     |████████████████████████████████| 79 kB 9.1 MB/s  eta 0:00:01
  Created wheel for folium: filename=folium-0.5.0-py3-none-any.whl size=76240 sha256=051eee477960b8282850e5f00f397e2ab9d3d427de35a94ee37a1f30e96742a2
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/b2/2f/2c/109e446b990d663ea5ce9b078b5e7c1a9c45cca91f377080f8
Successfully built folium
Libraries imported.


In [8]:
!pip install geopy
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


## 1. Exploring only those boroughs that contain the word 'Toronto'

In [30]:
options = ['Downtown Toronto','East Toronto','West Toronto','Central Toronto']

toronto_data = TorontoPostalCodes[TorontoPostalCodes['Borough'].isin(options)].reset_index(drop=True)
toronto_data

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   
15        M4L      East Toronto   
16        M5L  Downtown Toronto   
17        M4M      East Toronto   
18        M4N   Central Toronto   
19        M5N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
24        M5R   Central Toronto   
25        M6R      West Toronto   
26        M4S   Central Toronto   
27        M5S  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
30        M5T  Downtown Toronto   
31        M4V   Central Toronto   
32        M5V  Downtown Toronto   
33        M4W  Downtown Toronto   
34        M5W  Downtown Toronto   
35        M4X  Downtown Toronto   
36        M5X  Downtown Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                        Neighbourhood  Latitude  Longitude  
0                           Regent Park, Harbourfront   43.6555   -79.3626  
1         Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
2                            Garden District, Ryerson   43.6572   -79.3783  
3                                      St. James Town   43.6513   -79.3756  
4                                         The Beaches   43.6784   -79.2941  
5                                         Berczy Park   43.6456   -79.3754  
6                                  Central Bay Street   43.6564   -79.3860  
7                                            Christie   43.6683   -79.4205  
8                            Richmond, Adelaide, King   43.6496   -79.3833  
9                        Dufferin, Dovercourt Village   43.6655   -79.4378  
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936  
11                           Little Portugal, Trinity   43.6480   -79.4177  
12                       The Danforth West, Riverdale   43.6803   -79.3538  
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823  
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301  
15                     India Bazaar, The Beaches West   43.6693   -79.3155  
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823  
17                                    Studio District   43.6561   -79.3406  
18                                      Lawrence Park   43.7301   -79.3935  
19                                           Roselawn   43.7113   -79.4195  
20                                   Davisville North   43.7135   -79.3887  
21    Forest Hill North & West, Forest Hill Road Park   43.6966   -79.4120  
22                      High Park, The Junction South   43.6605   -79.4633  
23                  North Toronto West, Lawrence Park   43.7143   -79.4065  
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035  
25                             Parkdale, Roncesvalles   43.6469   -79.4521  
26                                         Davisville   43.7020   -79.3853  
27                     University of Toronto, Harbord   43.6629   -79.3987  
28                                 Runnymede, Swansea   43.6512   -79.4828  
29                        Moore Park, Summerhill East   43.6899   -79.3853  
30          Kensington Market, Chinatown, Grange Park   43.6541   -79.3978  
31  Summerhill West, Rathnelly, South Hill, Forest...   43.6861   -79.4025  
32  CN Tower, King and Spadina, Railway Lands, Har...   43.64

In [31]:
# create map of Toronto city using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [11]:
CLIENT_ID = 'OCGNXAFARHWAGWZ4BPFTF22JZFP0PRAGPBVSBGNUPQRIWB30' # your Foursquare ID
CLIENT_SECRET = 'L1NTEIW4IOTRCRWH23QRHWKLKHXR41VZ5S0IWJW0SQISLN33' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: OCGNXAFARHWAGWZ4BPFTF22JZFP0PRAGPBVSBGNUPQRIWB30
CLIENT_SECRET:L1NTEIW4IOTRCRWH23QRHWKLKHXR41VZ5S0IWJW0SQISLN33


In [12]:
toronto_data.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

In [32]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6555, -79.3626.


In [14]:
latitude = 43.6555
longitude = -79.3626
radius = 500
LIMIT = 100

url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=OCGNXAFARHWAGWZ4BPFTF22JZFP0PRAGPBVSBGNUPQRIWB30&client_secret=L1NTEIW4IOTRCRWH23QRHWKLKHXR41VZ5S0IWJW0SQISLN33&ll=43.6555,-79.3626&v=20180605&radius=500&limit=100'

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6006c9dfca6cb15547e39b5b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 23,
  'suggestedBounds': {'ne': {'lat': 43.660000004500006,
    'lng': -79.3563918719477},
   'sw': {'lat': 43.6509999955, 'lng': -79.36880812805231}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'di

In [35]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [36]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name      categories        lat        lng
0        Roselle Desserts          Bakery  43.653447 -79.362017
1           Tandem Coffee     Coffee Shop  43.653559 -79.361809
2  Figs Breakfast & Lunch  Breakfast Spot  43.655675 -79.364503
3      Morning Glory Cafe  Breakfast Spot  43.653947 -79.361149
4         The Yoga Lounge     Yoga Studio  43.655515 -79.364955

In [37]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

23 venues were returned by Foursquare.


## 2. Exploring Neighborhoods in Troronto

In [38]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [39]:


toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [40]:
print(toronto_venues.shape)
toronto_venues.head()

(1523, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront                43.6555                -79.3626   
1  Regent Park, Harbourfront                43.6555                -79.3626   
2  Regent Park, Harbourfront                43.6555                -79.3626   
3  Regent Park, Harbourfront                43.6555                -79.3626   
4  Regent Park, Harbourfront                43.6555                -79.3626   

                    Venue  Venue Latitude  Venue Longitude  Venue Category  
0        Roselle Desserts       43.653447       -79.362017          Bakery  
1           Tandem Coffee       43.653559       -79.361809     Coffee Shop  
2  Figs Breakfast & Lunch       43.655675       -79.364503  Breakfast Spot  
3      Morning Glory Cafe       43.653947       -79.361149  Breakfast Spot  
4         The Yoga Lounge       43.655515       -79.364955     Yoga Studio

In [41]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            94   
Brockton, Parkdale Village, Exhibition Place                           38   
Business reply mail Processing Centre, South Ce...                     14   
CN Tower, King and Spadina, Railway Lands, Harb...                     58   
Central Bay Street                                                     58   
Christie                                                               11   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             22   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           17   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                       4   
High Park, The Junction South                                           2   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              49   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               47   
Moore Park, Summerhill East                                             5   
North Toronto West, Lawrence Park                                       4   
Parkdale, Roncesvalles                                                 46   
Queen's Park, Ontario Provincial Government                            29   
Regent Park, Harbourfront                                              23   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     48   
St. James Town                                                         88   
St. James Town, Cabbagetown                                            43   
Stn A PO Boxes                                                        100   
Studio District                                                         7   
Summerhill West, Rathnelly, South Hill, Forest ...                      6   
The Annex, North Midtown, Yorkville                                    25   
The Beaches                                                             8   
The Danforth West, Riverdale                                           36   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         27   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             94   
Brockton, Parkdale Village, Exhibition Place                            38   
Business reply mail Processing Centre, South Ce...                      14   
CN Tower, King and Spadina, Railway Lands, Harb...                      58   
Central Bay Street                                                      58   
Christie                                                                11   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              22   
Davisville North                    

In [42]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 215 uniques categories.


In [43]:
options = ['Neighborhood']

v_data = toronto_venues[toronto_venues['Venue Category'].isin(options)].reset_index(drop=True)
v_data

Neighborhood  Neighborhood Latitude  \
0                                   The Beaches                43.6784   
1                            Central Bay Street                43.6564   
2                      Richmond, Adelaide, King                43.6496   
3  Brockton, Parkdale Village, Exhibition Place                43.6383   
4                                Stn A PO Boxes                43.6437   

   Neighborhood Longitude             Venue  Venue Latitude  Venue Longitude  \
0                -79.2941     Upper Beaches       43.680563       -79.292869   
1                -79.3860  Downtown Toronto       43.653232       -79.385296   
2                -79.3833  Downtown Toronto       43.653232       -79.385296   
3                -79.4301          Parkdale       43.640524       -79.432200   
4                -79.3787      Harbourfront       43.639526       -79.380688   

  Venue Category  
0   Neighborhood  
1   Neighborhood  
2   Neighborhood  
3   Neighborhood  
4   Neighborhood

## 3. Analysing Each Neighborhood

In [54]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# drop neighborhood column, as it is found to be one of the venue categories
toronto_onehot = toronto_onehot.drop(columns='Neighborhood')

# copy the actual neighborhood column from venues dataframe to the first column of onehot encoded dataframe
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'])
toronto_onehot

Neighborhood  Accessories Store  \
0                             Regent Park, Harbourfront                  0   
1                             Regent Park, Harbourfront                  0   
2                             Regent Park, Harbourfront                  0   
3                             Regent Park, Harbourfront                  0   
4                             Regent Park, Harbourfront                  0   
...                                                 ...                ...   
1518  Business reply mail Processing Centre, South C...                  0   
1519  Business reply mail Processing Centre, South C...                  0   
1520  Business reply mail Processing Centre, South C...                  0   
1521  Business reply mail Processing Centre, South C...                  0   
1522  Business reply mail Processing Centre, South C...                  0   

      American Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   
...                   ...          ...         ...                  ...   
1518                    0            0           0                    0   
1519                    0            0           0                    0   
1520                    0            0           0                    0   
1521                    0            0           0                    0   
1522                    0            0           0                    0   

      Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop  ...  \
0                    0          0           0           0  ...   
1                    0          0           0           0  ...   
2                    0          0           0           0  ...   
3                    0          0           0           0  ...   
4                    0          0           0           0  ...   
...                ...        ...         ...         ...  ...   
1518                 0          0           0           0  ...   
1519                 0          0           0           0  ...   
1520                 0          0           0           0  ...   
1521                 0          0           0           0  ...   
1522                 0          0           0           0  ...   

      Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                    0      0              0                              0   
1                    0      0              0                              0   
2                    0      0              0                              0   
3                    0      0              0                              0   
4                    0      0              0                              0   
...                ...    ...            ...                            ...   
1518                 0      0              0                              0   
1519                 0      0              0                              0   
1520                 0      0              0                              0   
1521                 0      0              0                              0   
1522                 0      0              0                              0   

      Video Game Store  Vietnamese Restaurant  Whisky Bar  Wine Bar  \
0                    0                      0           0         0   
1                    0                      0           0         0   
2                    0                      0           0         0   
3                    0                      0           0         0   
4                    0                      0           0         0   
...                ...               

In [53]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot = toronto_onehot.drop(columns='Neighborhood')
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'])
toronto_onehot

Neighborhood  Accessories Store  \
0                             Regent Park, Harbourfront                  0   
1                             Regent Park, Harbourfront                  0   
2                             Regent Park, Harbourfront                  0   
3                             Regent Park, Harbourfront                  0   
4                             Regent Park, Harbourfront                  0   
...                                                 ...                ...   
1518  Business reply mail Processing Centre, South C...                  0   
1519  Business reply mail Processing Centre, South C...                  0   
1520  Business reply mail Processing Centre, South C...                  0   
1521  Business reply mail Processing Centre, South C...                  0   
1522  Business reply mail Processing Centre, South C...                  0   

      American Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   
...                   ...          ...         ...                  ...   
1518                    0            0           0                    0   
1519                    0            0           0                    0   
1520                    0            0           0                    0   
1521                    0            0           0                    0   
1522                    0            0           0                    0   

      Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop  ...  \
0                    0          0           0           0  ...   
1                    0          0           0           0  ...   
2                    0          0           0           0  ...   
3                    0          0           0           0  ...   
4                    0          0           0           0  ...   
...                ...        ...         ...         ...  ...   
1518                 0          0           0           0  ...   
1519                 0          0           0           0  ...   
1520                 0          0           0           0  ...   
1521                 0          0           0           0  ...   
1522                 0          0           0           0  ...   

      Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                    0      0              0                              0   
1                    0      0              0                              0   
2                    0      0              0                              0   
3                    0      0              0                              0   
4                    0      0              0                              0   
...                ...    ...            ...                            ...   
1518                 0      0              0                              0   
1519                 0      0              0                              0   
1520                 0      0              0                              0   
1521                 0      0              0                              0   
1522                 0      0              0                              0   

      Video Game Store  Vietnamese Restaurant  Whisky Bar  Wine Bar  \
0                    0                      0           0         0   
1                    0                      0           0         0   
2                    0                      0           0         0   
3                    0                      0           0         0   
4                    0                      0           0         0   
...                ...               

In [55]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.026316   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.000000   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    American Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0              0.010638     0.021277    0.000000             0.000000   
1              0.000000     0.026316    0.000000             0.026316   
2              0.000000     0.000000    0.000000             0.000000   
3              0.000000     0.000000    0.000000             0.000000   
4              0.000000     0.000000    0.017241             0.000000   
5              0.000000     0.000000    0.000000             0.000000   
6              0.013158     0.000000    0.000000             0.000000   
7              0.030000     0.010000    0.000000             0.000000   
8              0.000000     0.000000    0.000000             0.000000   
9              0.000000     0.000000    0.000000             0.000000   
10             0.000000     0.058824    0.000000             0.000000   
11             0.030000     0.010000    0.000000             0.000000   
12             0.000000     0.000000    0.000000   

In [56]:
toronto_grouped.shape

(39, 215)

In [57]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [67]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...            Restaurant   
3  CN Tower, King and Spadina, Railway Lands, Har...           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0                 Hotel      Seafood Restaurant                  Café   
1           Coffee Shop  Thrift / Vintage Store             Gift Shop   
2           Coffee Shop             Yoga Studio                  Bank   
3    Italian Restaurant    Gym / Fitness Center                  Café   
4                  Café          Sandwich Place           Pizza Place   

  5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
0                Bakery               Restaurant   Japanese Restaurant   
1        Breakfast Spot       Italian Restaurant   Japanese Restaurant   
2         Deli / Bodega           Breakfast Spot             Bookstore   
3                   Bar  New American Restaurant                  Bank   
4       Bubble Tea Shop               Restaurant    Italian Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Cocktail Bar              Beer Bar          Deli / Bodega  
1                    Brewery   Hawaiian Restaurant         Sandwich Place  
2           Sushi Restaurant   Japanese Restaurant     Italian Restaurant  
3                     Bakery                  Park      French Restaurant  
4  Middle Eastern Restaurant                   Spa         Breakfast Spot

## 4. Clustering Neighborhoods

In [68]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 3, 3, 3, 3, 0, 3, 3, 3, 3], dtype=int32)

In [69]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head()

PostalCode           Borough                                Neighbourhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.6555   -79.3626               3           Coffee Shop   
1   43.6641   -79.3889               3      Sushi Restaurant   
2   43.6572   -79.3783               3           Coffee Shop   
3   43.6513   -79.3756               3           Coffee Shop   
4   43.6784   -79.2941               3                   Pub   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0        Breakfast Spot                   Pub                 Beer Store   
1           Coffee Shop   Martial Arts School            Bubble Tea Shop   
2        Clothing Store        Cosmetics Shop  Middle Eastern Restaurant   
3                  Café    Seafood Restaurant         Italian Restaurant   
4     Health Food Store                 Trail                     Bakery   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Italian Restaurant     Electronics Store            Restaurant   
1      Ramen Restaurant           Escape Room         Burrito Place   
2   Japanese Restaurant                 Hotel                  Café   
3             Gastropub   American Restaurant          Cocktail Bar   
4           Coffee Shop             Gastropub   Dumpling Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                     Spa       Thai Restaurant                Theater  
1   Portuguese Restaurant       College Theater    Japanese Restaurant  
2  Furniture / Home Store  Fast Food Restaurant                  Diner  
3              Restaurant        Clothing Store         Farmers Market  
4       Fish & Chips Shop  Fast Food Restaurant         Farmers Market

In [72]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examining Clusters

#### Cluster 1

In [74]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
7                                            Christie                  Café   
10  Harbourfront East, Union Station, Toronto Islands       Harbor / Marina   
18                                      Lawrence Park    Photography Studio   
23                  North Toronto West, Lawrence Park            Playground   
29                        Moore Park, Summerhill East       Thai Restaurant   
33                                           Rosedale            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7          Grocery Store                  Park            Playground   
10                  Café                  Park           Music Venue   
18                  Park                Lawyer            Donut Shop   
23              Gym Pool                  Park                Garden   
29                  Park                   Gym         Grocery Store   
33                  Park         Grocery Store           Candy Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
7            Candy Store            Baby Store           Coffee Shop   
10           Yoga Studio   Dumpling Restaurant     Fish & Chips Shop   
18     Fish & Chips Shop  Fast Food Restaurant        Farmers Market   
23           Yoga Studio   Dumpling Restaurant     Fish & Chips Shop   
29                 Trail      Doner Restaurant  Fast Food Restaurant   
33           Yoga Studio   Dumpling Restaurant     Fish & Chips Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
7            Yoga Studio     Electronics Store      Fish & Chips Shop  
10  Fast Food Restaurant        Farmers Market     Falafel Restaurant  
18    Falafel Restaurant           Event Space   Ethiopian Restaurant  
23  Fast Food Restaurant        Farmers Market     Falafel Restaurant  
29        Farmers Market    Falafel Restaurant            Event Space  
33  Fast Food Restaurant        Farmers Market     Falafel Restaurant

#### Cluster 2

In [75]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
22  High Park, The Junction South                  Park   

                       2nd Most Common Venue 3rd Most Common Venue  \
22  Residential Building (Apartment / Condo)           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
22   Dumpling Restaurant           Fish Market     Fish & Chips Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
22  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   10th Most Common Venue  
22            Event Space

#### Cluster 3

In [76]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
21  Forest Hill North & West, Forest Hill Road Park          Home Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                  Park                 Trail     Electronics Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21           Yoga Studio   Dumpling Restaurant           Fish Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21     Fish & Chips Shop  Fast Food Restaurant         Farmers Market

#### Cluster 4

In [77]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
0                           Regent Park, Harbourfront           Coffee Shop   
1         Queen's Park, Ontario Provincial Government      Sushi Restaurant   
2                            Garden District, Ryerson           Coffee Shop   
3                                      St. James Town           Coffee Shop   
4                                         The Beaches                   Pub   
5                                         Berczy Park           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
8                            Richmond, Adelaide, King                  Café   
9                        Dufferin, Dovercourt Village                  Park   
11                           Little Portugal, Trinity                   Bar   
12                       The Danforth West, Riverdale      Greek Restaurant   
13           Toronto Dominion Centre, Design Exchange           Coffee Shop   
14       Brockton, Parkdale Village, Exhibition Place                  Café   
15                     India Bazaar, The Beaches West  Fast Food Restaurant   
16                     Commerce Court, Victoria Hotel           Coffee Shop   
17                                    Studio District         Garden Center   
20                                   Davisville North        Breakfast Spot   
24                The Annex, North Midtown, Yorkville                  Café   
25                             Parkdale, Roncesvalles           Coffee Shop   
26                                         Davisville    Italian Restaurant   
27                     University of Toronto, Harbord                  Café   
28                                 Runnymede, Swansea                  Café   
30          Kensington Market, Chinatown, Grange Park                  Café   
31  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
32  CN Tower, King and Spadina, Railway Lands, Har...           Coffee Shop   
34                                     Stn A PO Boxes           Coffee Shop   
35                        St. James Town, Cabbagetown           Coffee Shop   
36             First Canadian Place, Underground city           Coffee Shop   
37                               Church and Wellesley   Japanese Restaurant   
38  Business reply mail Processing Centre, South C...            Restaurant   

   2nd Most Common Venue          3rd Most Common Venue  \
0         Breakfast Spot                            Pub   
1            Coffee Shop            Martial Arts School   
2         Clothing Store                 Cosmetics Shop   
3                   Café             Seafood Restaurant   
4      Health Food Store                          Trail   
5                  Hotel             Seafood Restaurant   
6                   Café                 Sandwich Place   
8            Coffee Shop                          Hotel   
9                 Bakery                           Café   
11           Coffee Shop          Vietnamese Restaurant   
12        Ice Cream Shop                           Café   
13                 Hotel                           Café   
14           Coffee Shop         Thrift / Vintage Store   
15                   Gym             Italian Restaurant   
16                 Hotel                           Café   
17       Coworking Space                           Park   
20     Food & Drink Shop                           Park   
24        Sandwich Place                     Donut Shop   
25                Bakery    Eastern European Restaurant   
26          Dessert Shop                 Sandwich Place   
27             Bookstore            Japanese Restaurant   
28           Pizza Place                            Pub   
30           Coffee Shop  Vegetarian / Vegan Restaurant   
31    Light Rail Station                   Liquor Store   
32    Italian Restaurant           Gym / Fitness Center   
34            Restaurant                          Hotel 

#### Cluster 5

In [78]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
19      Roselawn           IT Services           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19   Dumpling Restaurant           Flower Shop           Fish Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19     Fish & Chips Shop  Fast Food Restaurant        Farmers Market   

   9th Most Common Venue 10th Most Common Venue  
19    Falafel Restaurant            Event Space

## 6. Observations (on Cluster Characteristics)

#### 1. Cluster-4 is the largest, with 30 neighbourhoods, that are predominently food zones (coffee shops, cafes, restaurants) among its top-5 venues
#### 2. Cluster-1 is the 2nd largest with 6 neighbourhoods, that are predominently play/wellness zones (playground, park, yoga studio)
#### 3. Cluster-2 is with just one neighbourhood, that has residential buildings
#### 4. Cluster-3 is also with just one neighbourhoof, that has home-service stores as its top-most venue
#### 5. Cluster-5 is also with just one neighbourhood, that has IT-services as its top-most venue

# End of Part-3 of 3